Hola **Roberto**!

Soy **Patricio Requena** 👋. Es un placer ser el revisor de tu proyecto el día de hoy!

Revisaré tu proyecto detenidamente con el objetivo de ayudarte a mejorar y perfeccionar tus habilidades. Durante mi revisión, identificaré áreas donde puedas hacer mejoras en tu código, señalando específicamente qué y cómo podrías ajustar para optimizar el rendimiento y la claridad de tu proyecto. Además, es importante para mí destacar los aspectos que has manejado excepcionalmente bien. Reconocer tus fortalezas te ayudará a entender qué técnicas y métodos están funcionando a tu favor y cómo puedes aplicarlos en futuras tareas. 

_**Recuerda que al final de este notebook encontrarás un comentario general de mi parte**_, empecemos!

Encontrarás mis comentarios dentro de cajas verdes, amarillas o rojas, ⚠️ **por favor, no muevas, modifiques o borres mis comentarios** ⚠️:


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class=“tocSkip”></a>
Si todo está perfecto.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class=“tocSkip”></a>
Si tu código está bien pero se puede mejorar o hay algún detalle que le hace falta.
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class=“tocSkip”></a>
Si de pronto hace falta algo o existe algún problema con tu código o conclusiones.
</div>

Puedes responderme de esta forma:
<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class=“tocSkip”></a>
    
</div>

In [2]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'practicum_student',         # nombre de usuario
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # contraseña
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # puerto de conexión
             'db': 'data-analyst-final-project-db'}          # nombre de la base de datos

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# Exploración de tablas de datos.

In [3]:
query = """
    SELECT *
    FROM books
"""

books = pd.io.sql.read_sql(query, con = engine)
books.head(10)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

In [4]:
query_au = """
    SELECT *
    FROM authors
    LIMIT 10
"""
authors = pd.io.sql.read_sql(query_au, con = engine)
authors.head(10)

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

In [5]:
query_pu = """
    SELECT *
    FROM publishers
"""
publishers = pd.io.sql.read_sql(query_pu, con = engine)
publishers.head(10)

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

In [6]:
query_ra = """
    SELECT *
    FROM ratings
"""
ratings = pd.io.sql.read_sql(query_ra, con = engine)
ratings.head(10)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

In [7]:
query_re = """
    SELECT * 
    FROM reviews;
"""
reviews = pd.io.sql.read_sql(query_re, con = engine)
reviews.head(10)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

<div class="alert alert-block alert-success">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Bien, siempre es bueno dar una breve revisión de las tablas con las que trabajarás
</div>

## Encuentra el número de libros publicados después del 1 de enero de 2000.

In [8]:
query_date = """
    SELECT COUNT(title) AS cnt
    FROM books
    WHERE publication_date::date > '2000-01-01';
"""
pd.io.sql.read_sql(query_date, con = engine)

cnt
0  819

Se han publicado 819 libros en este siglo.

<div class="alert alert-block alert-success">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Buen trabajo con el cálculo! Aplicaste el filtro correcto para obtener los datos
</div>

## Encuentra el número de reseñas de usuarios y la calificación promedio para cada libro.

In [33]:
query_rating = """
    SELECT 
        books.title AS title,
        AVG(ratings.rating) AS avg_rating,
        COUNT(ratings.rating) AS cnt_rating
    FROM
        books
        LEFT JOIN ratings ON books.book_id = ratings.book_id
    GROUP BY 
        books.title
    ORDER BY
        cnt_rating DESC;
"""
pd.io.sql.read_sql(query_rating, con = engine)

title  avg_rating  cnt_rating
0                              Twilight (Twilight  #1)    3.662500         160
1                  The Hobbit  or There and Back Again    4.125000          88
2                               The Catcher in the Rye    3.825581          86
3                 Angels & Demons (Robert Langdon  #1)    3.678571          84
4    Harry Potter and the Prisoner of Azkaban (Harr...    4.414634          82
..                                                 ...         ...         ...
994                                         Open House    2.500000           2
995                            Where Are the Children?    4.000000           2
996                                         Black Hole    4.500000           2
997                                     From a Buick 8    4.500000           2
998  Disney's Beauty and the Beast (A Little Golden...    4.000000           1

[999 rows x 3 columns]

La calificación promedio con pocas reseñas no es tan confiable.

Entre más calificaciones mejor y mas confiable es la calificación promedio.

## Identifica la editorial que ha publicado el mayor número de libros con más de 50 páginas (esto te ayudará a excluir folletos y publicaciones similares de tu análisis).

In [10]:
query_p50 = """
    SELECT 
        publishers.publisher AS publisher,
        COUNT(books.title) AS cnt_title
    FROM
        publishers
        LEFT JOIN books ON books.publisher_id = publishers.publisher_id
    WHERE 
        books.num_pages > 50
    GROUP BY 
        publishers.publisher
    ORDER BY
        cnt_title DESC
    LIMIT 1;    
"""
pd.io.sql.read_sql(query_p50, con = engine)

publisher  cnt_title
0  Penguin Books         42

Penguin Books es la editorial con mas libros publicados con 42.

<div class="alert alert-block alert-success">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Correcto, bien utilizado las funciones de agregación para obtener los resultados
</div>

## Identifica al autor que tiene la más alta calificación promedio del libro: mira solo los libros con al menos 50 calificaciones.

In [11]:
query_maxra = """
    SELECT 
        authors.author AS author,
        AVG(subquery.avg_rating) AS avg_rating
    FROM 
       authors
    JOIN (
        SELECT 
            books.author_id AS author_id,
            AVG(ratings.rating) AS avg_rating,
            COUNT(ratings.rating) AS cnt_rating
        FROM
            books
        JOIN ratings ON books.book_id = ratings.book_id
        GROUP BY 
            books.author_id
        ) AS subquery ON subquery.author_id = authors.author_id
    WHERE
        subquery.cnt_rating > 50
    GROUP BY
        authors.author
    ORDER BY
        avg_rating DESC
    LIMIT 1
"""
pd.io.sql.read_sql(query_maxra, con = engine)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.288462

J.K. Rowling y Mary GrandPré autora e ilustradora respectivamente de Harry Potter tiene la calificación mas alta con 4.29

<div class="alert alert-block alert-success">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Perfecto! Parece que los libros de Harry Potter son los que mejor puntaje tienen
</div>

## Encuentra el número promedio de reseñas de texto entre los usuarios que calificaron más de 50 libros.

In [32]:
query_us = """
    SELECT
        reviews.username,
        COUNT(reviews.text) AS cnt_reviews,
        AVG(COUNT(reviews.text)) OVER () AS promedio_general
    FROM 
        reviews
        JOIN (
            SELECT
                username,
                COUNT(rating) AS cnt_rating
            FROM 
                ratings
            GROUP BY
                username  
                ) AS subquery ON subquery.username = reviews.username
    WHERE
        subquery.cnt_rating > 50
    GROUP BY
        reviews.username    
"""
pd.io.sql.read_sql(query_us, con = engine)

username  cnt_reviews  promedio_general
0     sfitzgerald           28         24.333333
1  jennifermiller           25         24.333333
2          xdavis           18         24.333333
3          paul88           22         24.333333
4      martinadam           27         24.333333
5       richard89           26         24.333333

Los usuarios que han calificado mas de 50 libro hacen en promedio 24.33 reseñas. Podría decirse que reseñan la mitad de los libros que califican.

# Concluciones

Se han publicado 819 libros en este siglo.

La calificación promedio con pocas reseñas no es tan confiable. Entre más calificaciones mejor más confiable es la calificación promedio.

Penguin Books es la editorial con mas libros publicados con 42.

J.K. Rowling y Mary GrandPré autora e ilustradora respectivamente de Harry Potter tiene la calificación más alta con 4.29

Los usuarios que han calificado mas de 50 libro hacen en promedio 24.33 reseñas. Podría decirse que reseñan la mitad de los libros que califican.




<div class="alert alert-block alert-success">
<b>Comentario general (1ra Iteracion)</b> <a class=“tocSkip”></a>

Buen trabajo con las diferentes consultas realizadas, se obtuvo los resultados adecuados aplicando las operaciones correctas en cada una.

Cómo consejo adicional, debido a que SQL es un lenguaje que verás en tu día a día cómo Data Analyst te animo a seguir practicando con el mismo para que en las entrevistas técnicas destaques cómo profesional! Será muy común que cuando postules a un puesto laboral te pidan hacer una prueba de SQL, te recomiendo revisar plataformas cómo Leetcode ya que tienen ejercicios de SQL para prepararte para las entrevistas, particularmente SQL lo encuentras en este link: https://leetcode.com/problemset/database/
    
Saludos!
</div>